iceoryx_examples/ara_com/ara/runtime.hpp

```diff
@@ -0,0 +1,171 @@
+
+        m_discovery.findService(
+            serviceIdentifier,
+            instanceIdentifier,
+            iox::cxx::nullopt,
+            [&](auto& service) {
+                iceoryxServiceContainer.push_back({service.getEventIDString(), service.getInstanceIDString()});
+            },
+            iox::popo::MessagingPattern::PUB_SUB);
+
+        m_discovery.findService(
+            serviceIdentifier,
+            instanceIdentifier,
+            iox::cxx::nullopt,
+            [&](auto& service) {
+                iceoryxServiceContainer.push_back({service.getEventIDString(), service.getInstanceIDString()});
+            },
+            iox::popo::MessagingPattern::REQ_RES);
+
+        // We need to make sure that all three internal services representing 'MinimalSkeleton' are available
+        ara::com::ServiceHandleContainer<com::ProxyHandleType> araServiceContainer;
+        if (verifyThatServiceIsComplete(iceoryxServiceContainer))
+        {
+            araServiceContainer.push_back({serviceIdentifier, instanceIdentifier});
+        }
+
+        return araServiceContainer;
+    }
+
+    ara::com::FindServiceHandle StartFindService(ara::com::FindServiceHandler<com::ProxyHandleType> handler,
+                                                 ara::com::ServiceIdentifier& serviceIdentifier,
+                                                 ara::com::InstanceIdentifier& instanceIdentifier) noexcept
+    {
+        /// @todo #1332 Are duplicate entries allowed?
+        m_callbacks.push_back({handler, {serviceIdentifier, instanceIdentifier}});
+
+        if (m_callbacks.size() == 1)
+        {
+            auto invoker = iox::popo::createNotificationCallback(invokeCallback, *this);
+            m_listener.attachEvent(m_discovery, iox::runtime::ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED, invoker)
+                .or_else([](auto) {
+                    std::cerr << "unable to attach discovery" << std::endl;
+                    std::exit(EXIT_FAILURE);
+                });
+        }
+
+        return ara::com::FindServiceHandle({serviceIdentifier, instanceIdentifier});
+    }
+
+    void StopFindService(ara::com::FindServiceHandle handle) noexcept
+    {
+        auto iter = m_callbacks.begin();
+        for (; iter != m_callbacks.end(); iter++)
+        {
+            if (iter->second.GetServiceId() == handle.GetServiceId()
+                && iter->second.GetInstanceId() == handle.GetInstanceId())
+            {
+                break;
+            }
+        }
+        if (iter != m_callbacks.end())
+        {
+            m_callbacks.erase(iter);
+        }
+
+        if (m_callbacks.empty())
+        {
+            m_listener.detachEvent(m_discovery, iox::runtime::ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED);
+        }
+    }
+
+  private:
+    explicit Runtime() noexcept = default;
+
+    bool verifyThatServiceIsComplete(ara::com::ServiceHandleContainer<com::ProxyHandleType>& container)
```

Comment:
I think this does not belong in the runtime as it deals with a very specific service. I have to think about whether/how we can generalize this with what we have.

I wonder whether counting is sufficient, but if all events/fields/methods have distinct names it should be. One way would be that the service should provide a method which returns this count and this is what we check against (instead of the 3 here). This requires us to have access to this type here though and I do not see the object/type which allows this here.

Reply:
You're right, this is just a workaround for this brief example. That's the reason why I added the comment. Typically you would query here a database and find out if the containter containts everything which is represented by the AUTOSAR service.